tty/imx6ull-uart/imx6ull-uart.c

```diff
@@ -77,21 +95,143 @@ typedef struct {
 	volatile uint32_t *base;
 	uint32_t mode;
 	uint16_t dev_no;
+	uint8_t busy;
 
 	handle_t cond;
 	handle_t inth;
 	handle_t lock;
+	handle_t openclose_lock;
+
+	uint8_t int_flags;
+
+	struct {
+		uint32_t parity;
+		uint32_t frame;
+		uint32_t hw_overrun;
+		uint32_t sw_overrun;
+	} counters;
+
+	lf_fifo_t *rx_sw_fifo; /* unprocessed characters read from HW FIFO */
 
 	libtty_common_t tty_common;
 } uart_t;
 
 uart_t uart = { 0 };
 
+#define READER 1
+#define WRITER 2
+
 #define MODULE_CLK 20000000
 
 #define BUFSIZE 4096
 
-void uart_thr(void *arg)
+#define RX_SW_FIFO_SIZE 256
+
+/* interrupt flags */
+#define RX_DONE        (1 << 0)
+#define TX_DONE        (1 << 1)
+#define PARITY_ERR     (1 << 2)
+#define FRAME_ERR      (1 << 3)
+#define HW_OVERRUN_ERR (1 << 4)
+#define SW_OVERRUN_ERR (1 << 5)
+
+/* register flags */
+#define USR1_RRDY      (1 << 9)
+#define USR1_TRDY      (1 << 13)
+#define USR1_FRAMERR   (1 << 10)
+#define USR1_PARITYERR (1 << 15)
+#define USR2_RDR       (1 << 0)
+#define USR2_ORE       (1 << 1)
+#define UTS_TXFULL     (1 << 4)
+#define UTS_SOFTRST    (1 << 0)
+#define UCR1_UARTEN    (1 << 0)
+#define UCR1_RRDYEN    (1 << 9)
+#define UCR1_TRDYEN    (1 << 13)
+#define UCR2_SRST      (1 << 0)
+#define UCR2_RXEN      (1 << 1)
+#define UCR2_TXEN      (1 << 2)
+#define UCR2_WS        (1 << 5)
+#define UCR2_IRTS      (1 << 14)
+#define UCR3_RXDMUXSEL (1 << 2)
+#define UCR3_RI        (1 << 8)
+#define UCR3_DCD       (1 << 9)
+#define UCR3_DSR       (1 << 10)
+#define UCR3_FRAERREN  (1 << 11)
+#define UCR3_PARERREN  (1 << 12)
+#define UCR4_OREN      (1 << 1)
+
+
+static int flags_to_id(int flags)
+{
+	switch (flags & (O_RDONLY | O_WRONLY | O_RDWR)) {
+		case O_RDONLY:
+			return READER;
+		case O_WRONLY:
+			return WRITER;
+		case O_RDWR:
+			return (READER | WRITER);
+		default:
+			return 0;
+	}
+}
+
+
+static int uart_open(int flags)
+{
+	/* TODO: set PGID? */
+	int id = flags_to_id(flags);
+
+	if (id <= 0)
+		return -EACCES;
+
+	mutexLock(uart.openclose_lock);
+
+	if (uart.busy & id) {
```

Comment:
ok, side effect is that I can't open UART twice (not that I need to, just it's possible on linux)

Reply:
You mean two readers?

Reply:
Imho there is only a point in restricting readers. I see no problem with having multiple writers or reader and writer at the same time